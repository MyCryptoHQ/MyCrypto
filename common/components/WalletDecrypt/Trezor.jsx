// @flow
import './Trezor.scss';
import React, { Component } from 'react';
import translate from 'translations';
import TrezorConnect from 'vendor/trezor-connect';
import DeterministicWalletsModal from './DeterministicWalletsModal';
import TrezorWallet from 'libs/wallet/trezor';
import DPATHS from 'config/dpaths.js';
const DEFAULT_PATH = DPATHS.TREZOR[0].value;

type State = {
  publicKey: string,
  chainCode: string,
  dPath: string,
  error: ?string,
  isLoading: boolean
};

export default class TrezorDecrypt extends Component {
  props: { onUnlock: any => void };
  state: State = {
    publicKey: '',
    chainCode: '',
    dPath: DEFAULT_PATH,
    error: null,
    isLoading: false
  };

  _handlePathChange = (dPath: string) => {
    this._handleConnect(dPath);
  };

  _handleConnect = (dPath: string = this.state.dPath) => {
    this.setState({
      isLoading: true,
      error: null
    });

    TrezorConnect.getXPubKey(
      dPath,
      res => {
        if (res.success) {
          this.setState({
            dPath,
            publicKey: res.publicKey,
            chainCode: res.chainCode,
            isLoading: false
          });
        } else {
          this.setState({
            error: res.error,
            isLoading: false
          });
        }
      },
      '1.5.2'
    );
  };

  _handleCancel = () => {
    this.setState({
      publicKey: '',
      chainCode: '',
      dPath: DEFAULT_PATH
    });
  };

  _handleUnlock = (address: string, index: number) => {
    this.props.onUnlock(new TrezorWallet(address, this.state.dPath, index));
  };

  render() {
    const { dPath, publicKey, chainCode, error, isLoading } = this.state;
    const showErr = error ? 'is-showing' : '';

    return (
      <section className="TrezorDecrypt col-md-4 col-sm-6">
        <button
          className="TrezorDecrypt-decrypt btn btn-primary btn-lg"
          onClick={() => this._handleConnect()}
          disabled={isLoading}
        >
          {isLoading ? 'Unlocking...' : translate('ADD_Trezor_scan')}
        </button>

        <div className="TrezorDecrypt-help">
          Guide:{' '}
          <a
            href="https://blog.trezor.io/trezor-integration-with-myetherwallet-3e217a652e08"
            target="_blank"
            rel="noopener"
          >
            How to use TREZOR with MyEtherWallet
          </a>
        </div>

        <div className={`TrezorDecrypt-error alert alert-danger ${showErr}`}>
          {error || '-'}
        </div>

        <a
          className="TrezorDecrypt-buy btn btn-sm btn-default"
          href="https://trezor.io/?a=myetherwallet.com"
          target="_blank"
          rel="noopener"
        >
          {translate('Donâ€™t have a TREZOR? Order one now!')}
        </a>

        <DeterministicWalletsModal
          isOpen={!!publicKey && !!chainCode}
          publicKey={publicKey}
          chainCode={chainCode}
          dPath={dPath}
          dPaths={DPATHS.TREZOR}
          onCancel={this._handleCancel}
          onConfirmAddress={this._handleUnlock}
          onPathChange={this._handlePathChange}
          walletType={translate('x_Trezor', true)}
        />
      </section>
    );
  }
}
