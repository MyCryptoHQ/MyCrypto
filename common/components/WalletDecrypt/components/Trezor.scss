.TrezorDecrypt {
  text-align: center;

  &-help {
    margin-top: 10px;
    font-size: 13px;
  }

  &-error {
    opacity: 0;
    transition: none;

    &.is-showing {
      opacity: 1;
    }
  }

  &-buy {
    margin: 10px 0;
  }

  &-message {
    display: flex;
    justify-content: center;
    align-items: center;

    .Spinner {
      margin-right: 16px;
    }
  }
}
