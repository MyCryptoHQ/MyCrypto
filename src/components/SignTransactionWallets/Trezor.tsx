import React from 'react';

import { translateRaw } from '@translations';
import { ISignComponentProps } from '@types';

import HardwareSignTransaction from './Hardware';

export default function SignTransactionTrezor({
  senderAccount,
  rawTransaction,
  onSuccess
}: ISignComponentProps) {
  return (
    <HardwareSignTransaction
      signerDescription={translateRaw('SIGN_TX_TREZOR_DESCRIPTION')}
      walletIconType="trezor-icon-lg"
      senderAccount={senderAccount}
      rawTransaction={rawTransaction}
      onSuccess={onSuccess}
    />
  );
}
