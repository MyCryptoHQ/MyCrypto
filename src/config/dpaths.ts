import { DPath } from '@types';

const ETH_DEFAULT: DPath = {
  label: 'Default (ETH)',
  value: "m/44'/60'/0'/0"
};

const ETH_TREZOR: DPath = {
  label: 'TREZOR (ETH)',
  value: "m/44'/60'/0'/0"
};

const ETH_LEDGER: DPath = {
  label: 'Ledger (ETH)',
  value: "m/44'/60'/0'"
};

const ETH_LEDGER_LIVE: DPath = {
  label: 'Ledger Live (ETH)',
  value: "m/44'/60'/0'/0/0",
  isHardened: true,
  getIndex: (addressIndex): string => `m/44'/60'/${addressIndex}'/0/0`
};

const ETC_LEDGER: DPath = {
  label: 'Ledger (ETC)',
  value: "m/44'/60'/160720'/0'"
};

const ETC_TREZOR: DPath = {
  label: 'TREZOR (ETC)',
  value: "m/44'/61'/0'/0"
};

const ETH_TESTNET: DPath = {
  label: 'Testnet (ETH)',
  value: "m/44'/1'/0'/0"
};

const EXP_DEFAULT: DPath = {
  label: 'Default (EXP)',
  value: "m/44'/40'/0'/0"
};

const UBQ_DEFAULT: DPath = {
  label: 'Default (UBQ)',
  value: "m/44'/108'/0'/0"
};

const POA_DEFAULT: DPath = {
  label: 'Default (POA)',
  value: "m/44'/60'/0'/0"
};

const TOMO_DEFAULT: DPath = {
  label: 'Default (TOMO)',
  value: "m/44'/889'/0'/0"
};

const MUSIC_DEFAULT: DPath = {
  label: 'Default (MUSIC)',
  value: "m/44'/184'/0'/0"
};

const EGEM_DEFAULT: DPath = {
  label: 'Default (EGEM)',
  value: "m/44'/1987'/0'/0"
};

const CLO_DEFAULT: DPath = {
  label: 'Default (CLO)',
  value: "m/44'/820'/0'/0"
};

const ETH_SINGULAR: DPath = {
  label: 'SingularDTV',
  value: "m/0'/0'/0'"
};

const RSK_TESTNET: DPath = {
  label: 'Testnet (RSK)',
  value: "m/44'/37310'/0'/0"
};

const RSK_MAINNET: DPath = {
  label: 'Mainnet (RSK)',
  value: "m/44'/137'/0'/0"
};

const GO_DEFAULT: DPath = {
  label: 'Default (GO)',
  value: "m/44'/6060'/0'/0"
};

const ATH_DEFAULT: DPath = {
  label: 'Default (ATH)',
  value: "m/44'/1620'/0'/0"
};

const ETHO_DEFAULT: DPath = {
  label: 'Default (ETHO)',
  value: "m/44'/1313114'/0'/0"
};

const MIX_DEFAULT: DPath = {
  label: 'Default (MIX)',
  value: "m/44'/76'/0'/0"
};

const REOSC_DEFAULT: DPath = {
  label: 'Default (REOSC)',
  value: "m/44'/2894'/0'/0"
};

const ARTIS_SIGMA1: DPath = {
  label: 'Sigma1 (ATS)',
  value: "m/44'/60'/0'/0"
};

const ARTIS_TAU1: DPath = {
  label: 'Tau1 (ATS)',
  value: "m/44'/60'/0'/0"
};

const THUNDERCORE_DEFAULT: DPath = {
  label: 'Default (THUNDERCORE)',
  value: "m/44'/1001'/0'/0"
};

const ETI_DEFAULT: DPath = {
  label: 'Default (ETI)',
  value: "m/44'/464'/0'/0"
};
const WEB_DEFAULT: DPath = {
  label: 'Default (WEB)',
  value: "m/44'/227'/0'/0"
};

const METADIUM_DEFAULT: DPath = {
  label: 'Default (METADIUM)',
  value: "m/44'/916'/0'/0"
};

const DEXON_DEFAULT: DPath = {
  label: 'Default (DEXON)',
  value: "m/44'/237'/0'/0"
};

const ASK_DEFAULT: DPath = {
  label: 'Default (ASK)',
  value: "m/44'/2221'/0'/0"
};

const ASK_TREZOR: DPath = {
  label: 'TREZOR (ASK)',
  value: "m/44'/2221'/0'/0"
};

const AUX_DEFAULT: DPath = {
  label: 'Default (AUX)',
  value: "m/44'/344'/0'/0"
};

const ERE_DEFAULT: DPath = {
  label: 'Default (ERE)',
  value: "m/44'/466'/0'/0"
};

const VOLTA_DEFAULT: DPath = {
  label: 'Default (VOLTA)',
  value: "m/44'/73799'/0'/0"
};

const EWC_DEFAULT: DPath = {
  label: 'Default (EWC)',
  value: "m/44'/246'/0'/0"
};

const AVAX_DEFAULT: DPath = {
  label: 'Default (AVAX)',
  value: "m/44'/60'/0'/0"
};

export const DPathsList = {
  ETH_DEFAULT,
  ETH_TREZOR,
  ETH_LEDGER,
  ETH_LEDGER_LIVE,
  ETC_LEDGER,
  ETC_TREZOR,
  ETH_TESTNET,
  EXP_DEFAULT,
  UBQ_DEFAULT,
  POA_DEFAULT,
  TOMO_DEFAULT,
  MUSIC_DEFAULT,
  EGEM_DEFAULT,
  CLO_DEFAULT,
  RSK_MAINNET,
  RSK_TESTNET,
  GO_DEFAULT,
  ATH_DEFAULT,
  ETHO_DEFAULT,
  MIX_DEFAULT,
  REOSC_DEFAULT,
  ARTIS_SIGMA1,
  ARTIS_TAU1,
  THUNDERCORE_DEFAULT,
  ETI_DEFAULT,
  WEB_DEFAULT,
  METADIUM_DEFAULT,
  DEXON_DEFAULT,
  ASK_DEFAULT,
  ASK_TREZOR,
  AUX_DEFAULT,
  ETH_SINGULAR,
  ERE_DEFAULT,
  VOLTA_DEFAULT,
  EWC_DEFAULT,
  AVAX_DEFAULT
};

// PATHS TO BE INCLUDED REGARDLESS OF WALLET FORMAT
// const EXTRA_PATHS = [ETH_SINGULAR];

// Full length deterministic wallet paths from BIP44
// https://github.com/bitcoin/bips/blob/master/bip-0044.mediawiki
// normal path length is 4, ledger is the exception at 3

// m / purpose' / coin_type' / account' / change / address_index
//   |          |            |          |        |
//   | constant |   index    |  index   | 0 or 1 |
//   |__________|____________|__________|________|

// whitespace strings are evaluated the same way as nospace strings, except they allow optional spaces between each portion of the string
// ie. "m / 44' / 0' / 0'" is valid, "m / 4 4' / 0' / 0'" is invalid
export const dPathRegex = /m\/4[4,9]'\/[0-9]+'\/[0-9]+('+$|'+(\/[0-1]+$))/;
// const whitespaceDPathRegex = /m\s*\/\s*44'\s*\/\s*[0-9]+\'\s*\/\s*[0-9]+(\'+$|\'+\s*(\/\s*[0-1]+$))/;

/**
 * Due to limitations in the Ledger ETH application, only derivation paths starting with
 * `m/44'/60'` and `m/44'/1'` can be checked.
 */
export const LEDGER_DERIVATION_PATHS: DPath[] = [
  ETH_DEFAULT,
  ETH_LEDGER,
  ETC_LEDGER,
  ETH_TESTNET,
  ETH_LEDGER_LIVE
];

/**
 * While Trezor does support hardened paths, it'd be very tedious for the user to check all the
 * paths currently, since the user would have to confirm each address individually.
 */
export const TREZOR_DERIVATION_PATHS: DPath[] = [
  ...Object.values(DPathsList).filter((path) => !path.isHardened)
];
