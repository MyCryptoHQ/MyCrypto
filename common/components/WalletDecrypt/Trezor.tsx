import DPATHS from 'config/dpaths';
import TrezorWallet from 'libs/wallet/trezor';
import React, { Component } from 'react';
import translate, { translateRaw } from 'translations';
import TrezorConnect from 'vendor/trezor-connect';
import DeterministicWalletsModal from './DeterministicWalletsModal';
import './Trezor.scss';
const DEFAULT_PATH = DPATHS.TREZOR[0].value;

interface Props {
  onUnlock(param: any): void;
}
interface State {
  publicKey: string;
  chainCode: string;
  dPath: string;
  error: string | null;
  isLoading: boolean;
}

export default class TrezorDecrypt extends Component<Props, State> {
  public state: State = {
    publicKey: '',
    chainCode: '',
    dPath: DEFAULT_PATH,
    error: null,
    isLoading: false
  };

  public render() {
    const { dPath, publicKey, chainCode, error, isLoading } = this.state;
    const showErr = error ? 'is-showing' : '';

    return (
      <section className="TrezorDecrypt col-md-4 col-sm-6">
        <button
          className="TrezorDecrypt-decrypt btn btn-primary btn-lg"
          onClick={this.handleNullConnect}
          disabled={isLoading}
        >
          {isLoading ? 'Unlocking...' : translate('ADD_Trezor_scan')}
        </button>

        <div className="TrezorDecrypt-help">
          Guide:{' '}
          <a
            href="https://blog.trezor.io/trezor-integration-with-myetherwallet-3e217a652e08"
            target="_blank"
            rel="noopener"
          >
            How to use TREZOR with MyEtherWallet
          </a>
        </div>

        <div className={`TrezorDecrypt-error alert alert-danger ${showErr}`}>
          {error || '-'}
        </div>

        <a
          className="TrezorDecrypt-buy btn btn-sm btn-default"
          href="https://trezor.io/?a=myetherwallet.com"
          target="_blank"
          rel="noopener"
        >
          {translate('Donâ€™t have a TREZOR? Order one now!')}
        </a>

        <DeterministicWalletsModal
          isOpen={!!publicKey && !!chainCode}
          publicKey={publicKey}
          chainCode={chainCode}
          dPath={dPath}
          dPaths={DPATHS.TREZOR}
          onCancel={this.handleCancel}
          onConfirmAddress={this.handleUnlock}
          onPathChange={this.handlePathChange}
          walletType={translateRaw('x_Trezor')}
        />
      </section>
    );
  }

  private handlePathChange = (dPath: string) => {
    this.setState({ dPath });
    this.handleConnect(dPath);
  };

  private handleConnect = (dPath: string = this.state.dPath): void => {
    this.setState({
      isLoading: true,
      error: null
    });

    // TODO: type vendor file
    (TrezorConnect as any).getXPubKey(
      dPath,
      res => {
        if (res.success) {
          this.setState({
            dPath,
            publicKey: res.publicKey,
            chainCode: res.chainCode,
            isLoading: false
          });
        } else {
          this.setState({
            error: res.error,
            isLoading: false
          });
        }
      },
      '1.5.2'
    );
  };

  private handleCancel = () => {
    this.setState({
      publicKey: '',
      chainCode: '',
      dPath: DEFAULT_PATH
    });
  };

  private handleUnlock = (address: string, index: number) => {
    this.props.onUnlock(new TrezorWallet(address, this.state.dPath, index));
  };

  private handleNullConnect = (): void => this.handleConnect();
}
