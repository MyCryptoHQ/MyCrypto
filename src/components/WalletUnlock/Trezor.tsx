import React, { PureComponent } from 'react';

import { Button } from '@mycrypto/ui';

import ConnectTrezor from '@assets/images/icn-connect-trezor-new.svg';
import { Box, BusyBottom, Heading, Spinner } from '@components';
import { getDPath, getDPaths, INetworkContext, useNetworks } from '@services';
import { ChainCodeResponse, WalletFactory } from '@services/WalletService';
import translate, { translateRaw } from '@translations';
import { BusyBottomConfig, DPath, FormData, TAddress, WalletId } from '@types';
import { withHook } from '@utils';

import DeterministicWallets from './DeterministicWallets';
import './Trezor.scss';
import UnsupportedNetwork from './UnsupportedNetwork';

//@todo: conflicts with comment in walletDecrypt -> onUnlock method
interface OwnProps {
  formData: FormData;
  onUnlock(param: any): void;
}

// @todo: nearly duplicates ledger component props
interface State {
  publicKey: string;
  chainCode: string;
  dPath: DPath;
  error: string | null;
  isLoading: boolean;
}

const WalletService = WalletFactory[WalletId.TREZOR];

class TrezorDecryptClass extends PureComponent<OwnProps & INetworkContext, State> {
  public state: State = {
    publicKey: '',
    chainCode: '',
    dPath:
      getDPath(this.props.getNetworkById(this.props.formData.network), WalletId.TREZOR) ||
      getDPaths(this.props.networks, WalletId.TREZOR)[0],
    error: null,
    isLoading: false
  };

  public render() {
    const { dPath, publicKey, chainCode, isLoading } = this.state;
    const networks = this.props.networks;
    const network = this.props.getNetworkById(this.props.formData.network);

    if (!dPath) {
      return <UnsupportedNetwork walletType={translateRaw('x_Trezor')} network={network} />;
    }

    if (publicKey && chainCode) {
      return (
        <div className="Mnemonic-dpath">
          <DeterministicWallets
            network={network}
            publicKey={publicKey}
            chainCode={chainCode}
            dPath={dPath}
            dPaths={getDPaths(networks, WalletId.TREZOR)}
            onCancel={this.handleCancel}
            onConfirmAddress={this.handleUnlock}
            onPathChange={this.handlePathChange}
          />
        </div>
      );
    } else {
      return (
        <Box p="2.5em">
          <Heading fontSize="32px" textAlign="center" fontWeight="bold">
            {translate('UNLOCK_WALLET')}{' '}
            {translateRaw('YOUR_WALLET_TYPE', { $walletType: translateRaw('X_TREZOR') })}
          </Heading>
          <div className="TrezorDecrypt">
            <div className="TrezorDecrypt-description">
              {translate('TREZOR_TIP')}
              <div className="TrezorDecrypt-img">
                <img src={ConnectTrezor} />
              </div>
            </div>
            {/* <div className={`TrezorDecrypt-error alert alert-danger ${showErr}`}>
              {error || '-'}
            </div> */}

            {isLoading ? (
              <div className="TrezorDecrypt-loading">
                <Spinner /> {translate('WALLET_UNLOCKING')}
              </div>
            ) : (
              <Button
                className="TrezorDecrypt-button"
                onClick={this.handleNullConnect}
                disabled={isLoading}
              >
                {translate('ADD_TREZOR_SCAN')}
              </Button>
            )}
            <div className="TrezorDecrypt-footer">
              <BusyBottom type={BusyBottomConfig.TREZOR} />
            </div>
          </div>
        </Box>
      );
    }
  }

  private handlePathChange = (dPath: DPath) => {
    this.setState({ dPath });
    this.handleConnect(dPath);
  };

  private handleConnect = (dPath: DPath): void => {
    this.setState({
      isLoading: true,
      error: null
    });

    WalletService.getChainCode(dPath.value)
      .then((res: ChainCodeResponse) => {
        this.setState({
          dPath,
          publicKey: res.publicKey,
          chainCode: res.chainCode,
          isLoading: false
        });
      })
      .catch((err: any) => {
        this.setState({
          error: err.message,
          isLoading: false
        });
      });
  };

  private handleCancel = () => {
    this.reset();
  };

  private handleUnlock = (address: TAddress, index: number) => {
    this.props.onUnlock(WalletService.init({ address, dPath: this.state.dPath.value, index }));
  };

  private handleNullConnect = (): void => {
    this.handleConnect(this.state.dPath);
  };

  private reset() {
    const networks = this.props.networks;
    const network = this.props.getNetworkById(this.props.formData.network);
    this.setState({
      publicKey: '',
      chainCode: '',
      dPath: getDPath(network, WalletId.TREZOR) || getDPaths(networks, WalletId.TREZOR)[0]
    });
  }
}

export const TrezorDecrypt = withHook(useNetworks)(TrezorDecryptClass);
