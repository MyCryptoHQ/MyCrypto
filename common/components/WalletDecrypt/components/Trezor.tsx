import { TrezorWallet, TREZOR_MINIMUM_FIRMWARE } from 'libs/wallet';
import React, { PureComponent } from 'react';
import translate, { translateRaw } from 'translations';
import TrezorConnect from 'vendor/trezor-connect';
import DeterministicWalletsModal from './DeterministicWalletsModal';
import './Trezor.scss';
import { Spinner } from 'components/ui';
import { getNetworkConfig } from 'selectors/config';
import { AppState } from 'reducers';
import { connect } from 'react-redux';
import { SecureWalletName } from 'config';
import { DPath } from 'config/dpaths';
import { getPaths, getSingleDPath } from 'utils/network';

//todo: conflicts with comment in walletDecrypt -> onUnlock method
interface OwnProps {
  onUnlock(param: any): void;
}

interface StateProps {
  dPath: DPath;
}

// todo: nearly duplicates ledger component props
interface State {
  publicKey: string;
  chainCode: string;
  dPath: string;
  error: string | null;
  isLoading: boolean;
}

type Props = OwnProps & StateProps;

class TrezorDecryptClass extends PureComponent<Props, State> {
  public state: State = {
    publicKey: '',
    chainCode: '',
    dPath: this.props.dPath.value,
    error: null,
    isLoading: false
  };

  public render() {
    const { dPath, publicKey, chainCode, error, isLoading } = this.state;
    const showErr = error ? 'is-showing' : '';

    return (
      <div className="TrezorDecrypt">
        <button
          className="TrezorDecrypt-decrypt btn btn-primary btn-lg btn-block"
          onClick={this.handleNullConnect}
          disabled={isLoading}
        >
          {isLoading ? (
            <div className="TrezorDecrypt-message">
              <Spinner light={true} />
              Unlocking...
            </div>
          ) : (
            translate('ADD_Trezor_scan')
          )}
        </button>

        <a
          className="TrezorDecrypt-buy btn btn-sm btn-default"
          href="https://trezor.io/?a=myetherwallet.com"
          target="_blank"
          rel="noopener noreferrer"
        >
          {translate('Donâ€™t have a TREZOR? Order one now!')}
        </a>

        <div className={`TrezorDecrypt-error alert alert-danger ${showErr}`}>{error || '-'}</div>

        <div className="TrezorDecrypt-help">
          Guide:{' '}
          <a
            href="https://blog.trezor.io/trezor-integration-with-myetherwallet-3e217a652e08"
            target="_blank"
            rel="noopener noreferrer"
          >
            How to use TREZOR with MyEtherWallet
          </a>
        </div>

        <DeterministicWalletsModal
          isOpen={!!publicKey && !!chainCode}
          publicKey={publicKey}
          chainCode={chainCode}
          dPath={dPath}
          dPaths={getPaths(SecureWalletName.TREZOR)}
          onCancel={this.handleCancel}
          onConfirmAddress={this.handleUnlock}
          onPathChange={this.handlePathChange}
          walletType={translateRaw('x_Trezor')}
        />
      </div>
    );
  }

  private handlePathChange = (dPath: string) => {
    this.setState({ dPath });
    this.handleConnect(dPath);
  };

  private handleConnect = (dPath: string = this.state.dPath): void => {
    this.setState({
      isLoading: true,
      error: null
    });

    (TrezorConnect as any).getXPubKey(
      dPath,
      res => {
        if (res.success) {
          this.setState({
            dPath,
            publicKey: res.publicKey,
            chainCode: res.chainCode,
            isLoading: false
          });
        } else {
          this.setState({
            error: res.error,
            isLoading: false
          });
        }
      },
      TREZOR_MINIMUM_FIRMWARE
    );
  };

  private handleCancel = () => {
    this.reset();
  };

  private handleUnlock = (address: string, index: number) => {
    this.props.onUnlock(new TrezorWallet(address, this.state.dPath, index));
    this.reset();
  };

  private handleNullConnect = (): void => this.handleConnect();

  private reset() {
    this.setState({
      publicKey: '',
      chainCode: '',
      dPath: this.props.dPath.value
    });
  }
}

function mapStateToProps(state: AppState): StateProps {
  const network = getNetworkConfig(state);
  return {
    dPath: getSingleDPath(SecureWalletName.TREZOR, network)
  };
}

export const TrezorDecrypt = connect(mapStateToProps)(TrezorDecryptClass);
