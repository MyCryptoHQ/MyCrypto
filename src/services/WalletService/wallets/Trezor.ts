import TrezorConnect from 'trezor-connect';

import { DPathsList, TREZOR_DERIVATION_PATHS } from '@config/dpaths';
import { DPath, WalletId } from '@types';

import HardwareWallet, { KeyInfo } from './HardwareWallet';
import { getFullPath } from './helpers';

export default class Trezor extends HardwareWallet {
  private cache: { [key: string]: KeyInfo } = {};

  public async initialize(): Promise<void> {
    TrezorConnect.manifest({
      email: 'support@mycrypto.com',
      appUrl: 'https://beta.mycrypto.com'
    });

    this.cache = {};

    // Fetch a random address to ensure the connection works
    try {
      await this.getAddress(DPathsList.ETH_DEFAULT, 50);
    } catch (err) {
      console.debug('[Trezor]: Error connecting to device');
      throw err;
    }
  }

  public async prefetch(paths: DPath[]): Promise<{ [key: string]: KeyInfo }> {
    const bundle = paths.filter((path) => !path.isHardened).map((path) => ({ path: path.value }));

    const response: any = await TrezorConnect.getPublicKey({ bundle }); // @todo - figure out these types

    for (const { serializedPath, chainCode, publicKey } of response.payload) {
      this.cache[serializedPath] = { chainCode, publicKey };
    }

    return this.cache;
  }

  public getDPaths(): DPath[] {
    return TREZOR_DERIVATION_PATHS;
  }

  protected getWalletType(): WalletId.TREZOR_NEW {
    return WalletId.TREZOR_NEW;
  }

  protected async getKeyInfo(path: DPath): Promise<KeyInfo> {
    if (this.cache[path.value]) {
      return this.cache[path.value];
    }

    const response = await TrezorConnect.getPublicKey({ path: path.value });

    return {
      publicKey: response.payload.publicKey,
      chainCode: response.payload.chainCode
    };
  }

  protected async getHardenedAddress(path: DPath, index: number): Promise<string> {
    /**
     * @todo: Add support for getting multiple addresses at the same time. For reference:
     * https://github.com/trezor/connect/blob/develop/docs/methods/ethereumGetAddress.md
     */
    const response = await TrezorConnect.ethereumGetAddress({ path: getFullPath(path, index) });

    return response.payload.address;
  }
}
