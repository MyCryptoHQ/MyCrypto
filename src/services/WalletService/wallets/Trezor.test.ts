import TrezorConnect from 'trezor-connect';

import { DPathsList } from '@config/dpaths';

import Trezor from './Trezor';

jest.mock('trezor-connect');

describe('Trezor', () => {
  it('generates an address from a derivation path', async () => {
    const wallet = new Trezor();
    await wallet.initialize();

    expect(TrezorConnect.manifest).toHaveBeenCalledTimes(1);

    await expect(wallet.getAddress(DPathsList.ETH_DEFAULT, 10)).resolves.toMatchSnapshot();
    await expect(wallet.getAddress(DPathsList.ETH_DEFAULT, 15)).resolves.toMatchSnapshot();
  });

  it('generates an address from a hardened derivation path', async () => {
    const wallet = new Trezor();
    await wallet.initialize();

    expect(TrezorConnect.manifest).toHaveBeenCalledTimes(2);

    await expect(wallet.getAddress(DPathsList.ETH_LEDGER_LIVE, 10)).resolves.toMatchSnapshot();
    await expect(wallet.getAddress(DPathsList.ETH_LEDGER_LIVE, 15)).resolves.toMatchSnapshot();
  });

  it('pre-fetches multiple addresses', async () => {
    const wallet = new Trezor();
    await wallet.initialize();

    const paths = [DPathsList.ETH_DEFAULT, DPathsList.ETH_TESTNET];

    await expect(wallet.prefetch(paths)).resolves.toMatchSnapshot();
  });
});
