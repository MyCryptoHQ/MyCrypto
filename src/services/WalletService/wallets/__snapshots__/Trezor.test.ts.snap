// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Trezor generates an address from a derivation path 1`] = `
Object {
  "address": "0x64E8A4EbF08cbAb754F9BEcBe80FA3CDFfbbdb1b",
  "path": "m/44'/60'/0'/0/10",
  "type": "TREZOR_NEW",
}
`;

exports[`Trezor generates an address from a derivation path 2`] = `
Object {
  "address": "0xb3363c233026E498c0a651D5E577396eCAca4AcB",
  "path": "m/44'/60'/0'/0/15",
  "type": "TREZOR_NEW",
}
`;

exports[`Trezor generates an address from a hardened derivation path 1`] = `
Object {
  "address": "0x0719f46C96047A4f8A791394E338c8108E56F246",
  "path": "m/44'/60'/10'/0/0",
  "type": "TREZOR_NEW",
}
`;

exports[`Trezor generates an address from a hardened derivation path 2`] = `
Object {
  "address": "0xA900c74246933B5F447C96581C5554Ac1bf36A92",
  "path": "m/44'/60'/15'/0/0",
  "type": "TREZOR_NEW",
}
`;

exports[`Trezor pre-fetches multiple addresses 1`] = `
Object {
  "m/44'/1'/0'/0": Object {
    "chainCode": "9d73a228d784f361eed3910b1d49750b33bc8aea09180b78abb71a09a17ae689",
    "publicKey": "02bc7ab0a01997363ba548279abf8302ecc50ed376fe74ba3133f1678346ce0c5d",
  },
  "m/44'/60'/0'/0": Object {
    "chainCode": "51f696d1838ec2986b979577cc43c1098e26fe34d9abaf319df00a7eb20e0311",
    "publicKey": "021c3e866ccb8f158431f5036319dc16f0409bd385d796fbc122a22819f0ec9017",
  },
}
`;
